components/automate-cli/cmd/chef-automate/restart.go

```diff
@@ -49,6 +123,166 @@ func runRestartServices(cmd *cobra.Command, args []string) error {
+		errChan <- &ResultWithError{}
+	}
+}
+
+func runRestartCmdForBackend(infra *AutomateHAInfraDetails, flags *RestartCmdFlags, rs restartFromBastion, mutex *sync.Mutex, errChan chan *ResultWithError) {
+	if flags.postgresql {
+		flags.automate = false
+		flags.chefServer = false
+		restartOnGivenNode(flags, POSTGRESQL, infra, rs, mutex, errChan)
+	} else {
+		errChan <- &ResultWithError{}
+	}
+
+	if flags.opensearch {
+		flags.automate = false
+		flags.chefServer = false
+		flags.postgresql = false
+		restartOnGivenNode(flags, OPENSEARCH, infra, rs, mutex, errChan)
+	} else {
+		errChan <- &ResultWithError{}
+	}
+}
+
+func restartOnGivenNode(flags *RestartCmdFlags, nodeType string, infra *AutomateHAInfraDetails, rs restartFromBastion, mutex *sync.Mutex, resultErrChan chan *ResultWithError) {
+	go func(flags RestartCmdFlags, resultErrChan chan *ResultWithError) {
+		writer := cli.NewWriter(os.Stdout, os.Stderr, os.Stdin)
+		sshUtil := NewSSHUtil(&SSHConfig{})
+		nodeMap := constructNodeMapForAllNodeTypes(&flags, infra)
+		cmdResult, err := rs.executeRemoteExecutor(nodeMap, sshUtil, writer)
+		ResultWithError := &ResultWithError{result:cmdResult,err:err }
+		resultErrChan <-ResultWithError
+	}(*flags, resultErrChan)
+}
+
+func getChannelValue(errChan chan *ResultWithError) error {
+	for i := 0; i < 4; i++ {
+		errVal := <-errChan
+		if errVal.err != nil {
+			return errVal.err
+		}
+	}
+	return nil
+}
+
+func handleManagedServices(flags *RestartCmdFlags) error {
+	if flags.postgresql && flags.opensearch {
+		return nil
+	}
+	if flags.postgresql {
+		return status.Errorf(status.InvalidCommandArgsError, ERROR_ON_MANAGED_SERVICES, POSTGRESQL)
+	}
+	if flags.opensearch {
+		return status.Errorf(status.InvalidCommandArgsError, ERROR_ON_MANAGED_SERVICES, OPENSEARCH)
+	}
+	return nil
+}
+
+func constructNodeMapForAllNodeTypes(flags *RestartCmdFlags, infra *AutomateHAInfraDetails) *NodeTypeAndCmd {
+	nodeMap := &NodeTypeAndCmd{
+		Frontend: &Cmd{
+			CmdInputs: &CmdInputs{
+				Cmd:                      RESTART_FRONTEND_COMMAND,
+				Single:                   false,
+				NodeType:                 false,
+				HideSSHConnectionMessage: true,
+			},
+		},
+		Automate: &Cmd{
+			CmdInputs: &CmdInputs{
+				Cmd:                      RESTART_FRONTEND_COMMAND,
+				Single:                   false,
+				ErrorCheckEnableInOutput: true,
+				WaitTimeout:              int(flags.timeout),
+				NodeIps:                  []string{flags.node},
+				NodeType:                 flags.automate,
+				HideSSHConnectionMessage: true,
+			},
+		},
+		ChefServer: &Cmd{
+			CmdInputs: &CmdInputs{
+				Cmd:                      RESTART_FRONTEND_COMMAND,
+				WaitTimeout:              int(flags.timeout),
+				Single:                   false,
+				ErrorCheckEnableInOutput: true,
+				NodeIps:                  []string{flags.node},
+				NodeType:                 flags.chefServer,
+				HideSSHConnectionMessage: true,
+			},
+		},
+		Postgresql: &Cmd{
+			CmdInputs: &CmdInputs{
```

Comment:
if user is giving timeout value pass it to the util for pg and os also

Reply:
done